File: src/main/session-clock.ts
export class SessionClock {
  private baseline = performance.now();
  private eligible = false;

  constructor(private countedMs: number, private paused: boolean) {}

  setPaused(paused: boolean): void {
    this.update();
    this.paused = paused;
    this.baseline = performance.now();
  }

  setEligible(eligible: boolean): void {
    this.update();
    this.eligible = eligible;
  }

  update(): number {
    const now = performance.now();
    const elapsed = now - this.baseline;
    this.baseline = now;
    // Unknown long gaps are never credited; normal renderer stalls remain harmless.
    if (!this.paused && this.eligible && elapsed >= 0 && elapsed < 10_000) this.countedMs += elapsed;
    return this.countedMs;
  }
}

File: Sources/TiddlyCore/SessionClock.swift
import Foundation

/// Counts time only while the agent app is frontmost and the pet is not paused. Gaps of 10 seconds or more are not credited.
public final class SessionClock {
  public static let maxCreditedGap: TimeInterval = 10
  private let now: () -> TimeInterval
  private var baseline: TimeInterval
  private var eligible = false
  private var paused: Bool
  private var countedMs: Double

  public init(countedMs: Double, paused: Bool, now: @escaping () -> TimeInterval = { ProcessInfo.processInfo.systemUptime }) {
    self.now = now; self.countedMs = countedMs; self.paused = paused; baseline = now()
  }

  public func setPaused(_ paused: Bool) { _ = update(); self.paused = paused; baseline = now() }
  public func setEligible(_ eligible: Bool) { _ = update(); self.eligible = eligible }

  @discardableResult
  public func update() -> Double {
    let current = now()
    let elapsed = current - baseline
    baseline = current
    if !paused && eligible && elapsed >= 0 && elapsed < SessionClock.maxCreditedGap { countedMs += elapsed * 1000 }
    return countedMs
  }
}
